Add `python.interpreters.include` and `python.interpreters.exclude` settings

sharon-wang opened the pull request:
### Summary
- addresses #3574 for both the JS and native locators
- also implements #6254 for the JS locator (previously [only implemented for the native locator](https://github.com/posit-dev/positron/pull/6261))
- introduces new settings
    - `python.interpreters.include`: List of folders to search for Python installations. These folders are searched in addition to the default folders for your operating system.
    - `python.interpreters.exclude`: List of interpreter paths or folders to exclude from the available Python installations. These interpreters will not be displayed in the Positron UI. This option takes precedence over the `include` option.
- adds new command available via Command Palette: `python.interpreters.debugInfo`: "Print interpreter debug information to Output"
    - Opens the Python Language Pack Output and prints python interpreter settings info as well as info about each discovered python runtime

#### Settings UI

<img width="1020" alt="image" src="https://github.com/user-attachments/assets/137d7798-50c8-4668-9444-db9917e91417" />

#### Settings JSON

```json
    "python.interpreters.include": [
        "~/scratch"
    ],
    "python.interpreters.exclude": [
        "/usr/bin/python3",
        "/opt/homebrew",
        "/usr/local/bin",
        "/opt/python",
        "~/scratch"
    ]
```

#### Python Language Pack Output
Some log prefixes/messages to search for:
- `pythonRuntimeDiscoverer`: to find out if any runtimes were filtered out during discovery
- `shouldIncludeInterpreter`: whether or not the interpreter will be included based on the user settings
- `Not registering runtime ${extraData.pythonPath} as it is excluded via user settings.`: if the runtime is not getting registered because the user excluded it
- `getAdditionalEnvDirs`: native python finder -- list all the additional directories being searched, such as the user-included dirs

<details><summary>Sample Output from `python.interpreters.debugInfo`: "Print interpreter debug information to Output"</summary>

```js
2025-02-24 10:43:17.026 [info] =====================================================================
2025-02-24 10:43:17.026 [info] =============== [START] PYTHON INTERPRETER DEBUG INFO ===============
2025-02-24 10:43:17.026 [info] =====================================================================
2025-02-24 10:43:17.026 [info] Python interpreter settings: {
  defaultInterpreterPath: '/Users/hello/qa-example-content/.venv/bin/python',
  'interpreters.include': [ '/Users/hello/scratch' ],
  'interpreters.exclude': [
    '/usr/bin/python3',
    '/opt/homebrew',
    '/usr/local/bin',
    '/opt/python',
    '/Users/hello/scratch'
  ]
}
2025-02-24 10:43:17.026 [info] Python interpreters discovered: [
  {
    name: 'Python 3.13.1 64-bit',
    path: '/opt/homebrew/bin/python3',
    versionInfo: { version: '3.13.1', supportedVersion: true },
    envInfo: { envType: 'Global', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: false,
      excludedInSettings: true
    }
  },
  {
    name: 'Python 3.10.4 64-bit',
    path: '/opt/python/3.10.4/bin/python',
    versionInfo: { version: '3.10.4', supportedVersion: true },
    envInfo: { envType: 'Unknown', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: false,
      excludedInSettings: true
    }
  },
  {
    name: "Python 3.12.2 64-bit ('3.12.2': pyenv)",
    path: '/Users/hello/.pyenv/versions/3.12.2/bin/python',
    versionInfo: { version: '3.12.2', supportedVersion: true },
    envInfo: { envType: 'Pyenv', envName: '3.12.2' },
    enablementInfo: {
      visibleInUI: true,
      includedInSettings: false,
      excludedInSettings: false
    }
  },
  {
    name: "Python 3.6.15 64-bit ('3.6.15': pyenv)",
    path: '/Users/hello/.pyenv/versions/3.6.15/bin/python',
    versionInfo: { version: '3.6.15', supportedVersion: false },
    envInfo: { envType: 'Pyenv', envName: '3.6.15' },
    enablementInfo: {
      visibleInUI: true,
      includedInSettings: false,
      excludedInSettings: false
    }
  },
  {
    name: "Python 3.10.4 ('.venv': venv)",
    path: '/Users/hello/qa-example-content/.venv/bin/python',
    versionInfo: { version: '3.10.4', supportedVersion: true },
    envInfo: { envType: 'Venv', envName: '.venv' },
    enablementInfo: {
      visibleInUI: true,
      includedInSettings: false,
      excludedInSettings: false
    }
  },
  {
    name: 'Python 3.10.4 64-bit (custom)',
    path: '/Users/hello/scratch/3.10.4/bin/python',
    versionInfo: { version: '3.10.4', supportedVersion: true },
    envInfo: { envType: 'Unknown', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: true,
      excludedInSettings: true
    }
  },
  {
    name: 'Python 3.9.6 64-bit',
    path: '/usr/bin/python3',
    versionInfo: { version: '3.9.6', supportedVersion: true },
    envInfo: { envType: 'Global', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: false,
      excludedInSettings: true
    }
  },
  {
    name: 'Python 3.10.4 64-bit',
    path: '/usr/local/bin/python3',
    versionInfo: { version: '3.10.4', supportedVersion: true },
    envInfo: { envType: 'Global', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: false,
      excludedInSettings: true
    }
  },
  {
    name: 'Python 3.11.6 64-bit',
    path: '/usr/local/bin/python3.11',
    versionInfo: { version: '3.11.6', supportedVersion: true },
    envInfo: { envType: 'Global', envName: '' },
    enablementInfo: {
      visibleInUI: false,
      includedInSettings: false,
      excludedInSettings: true
    }
  }
]
2025-02-24 10:43:17.026 [info] =====================================================================
2025-02-24 10:43:17.026 [info] ================ [END] PYTHON INTERPRETER DEBUG INFO ================
2025-02-24 10:43:17.026 [info] =====================================================================


```

</details> 


### Release Notes

#### New Features

- New settings to include additional Python interpreter search directories or exclude Python interpreters from the UI (#3574)

#### Bug Fixes

- filters out unsupported (<3.8) Python versions from the `Python: Select Interpreter` dropdown (#3740) @isabelizimm

### QA Notes

- Please verify these options using both the JS and Native locators. Toggle between the locators by updating setting `python.locator`.
- Positron will need to be restarted upon changing the `python.interpreters.include`/`python.interpreters.exclude` settings so that discovery can re-run with the settings applied
- If an interpreter path is captured in both options, it will be excluded
- If a user has a python install they want to include at path `/1/2/3/4/5/3.10.4/bin/python`, they should use `/1/2/3/4/5` or `/1/2/3/4/5/3.10.4` as the included directory.
- Windows users can specify paths with either path separator
- Relative paths specified in the options are ignored
- If a user starts an interpreter, then excludes or un-includes it and restarts Positron, it will still be selected upon restart
    - if they shut it down and then restart, it should no longer show

sharon-wang committed:
add option `python.interpreters.include`

This option allows user-specified Python search paths to be included as part of Python installation discovery. This is useful when the Python interpreter is not installed in a standard location.

The option is a list of unique paths to include in the search for Python interpreters. The paths can be absolute or relative to the user's home directory. The default value is an empty list.

sharon-wang committed:
add `python.interpreters.exclude` which is being recognized but not quite working yet

sharon-wang committed:
don't register user-excluded runtimes

sharon-wang committed:
add interpreters.include/exclude to python settings interfaces

sharon-wang committed:
add user-specified envs js locator for `python.interpreters.include` setting

sharon-wang committed:
add `/opt/python` path to js locator search dirs

sharon-wang committed:
add more detail to settings descriptions

sharon-wang committed:
move additional dirs code to avoid circular dependency

sharon-wang committed:
add more logging around additional dirs discovery

sharon-wang committed:
stub config apis for python runtime manager test

this is needed because the config api is now used to check for included/excluded runtimes as part of registering runtimes.

sharon-wang committed:
use `shouldIncludeInterpreter` prefix for all interpreter settings log messages

sharon-wang commented:
Opening this up for review, but I am working on adding a JS locator test for the new `UserSpecifiedEnvironmentLocator` that I'll push when done.

QUESTION: For the native locator, shall I just leave a note in `nativePythonFinder.unit.test.ts` to add 
testing for `python.interpreters.include` once we switch over? Currently the test is skipped and I'm not sure off the top of my head how to switch the locator for the unit tests.

melissa-barca commented on extensions/positron-python/src/client/positron/interpreterSettings.ts:
```diff
@@ -0,0 +1,86 @@
+import { INTERPRETERS_EXCLUDE_SETTING_KEY, INTERPRETERS_INCLUDE_SETTING_KEY } from '../common/constants';
+
+/**
+ * Gets the list of interpreters that the user has explicitly included in the settings. Converts
+ * relative and aliased paths to absolute paths.
+ * @returns List of interpreters that the user has explicitly included in the settings.
```
Where are we converting relative and aliases paths? It looks like we're just ignoring them.

sharon-wang replied:
oops good catch! I meant to resolve aliased paths and not include relative paths. I just pushed a commit to resolve aliased paths.

For relative paths, my thinking is that it's ambiguous where the paths are relative to. Are they relative to the home directory or to the project directory? Does it make sense for an admin to set a relative path as a default? I figured we can document that only absolute or absolute-but-aliased paths are valid here.

melissa-barca commented on extensions/positron-python/src/client/positron/interpreterSettings.ts:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * Check whether an interpreter should be included in the list of discovered interpreters.
+ * If an interpreter is both explicitly included and excluded, it will be included.
```
Are you confident this is the right approach? I'm wondering if we'd run into a situation where a Workbench admin has explicitly excluded an interpreter, but then a confused user tries to include it. We might want to consider implementing a way to log this use for admins, but we don't have monitoring implemented yet in Workbench for VS Code or Positron.

sharon-wang replied:
The language pack output will have a bit of info about why certain interpreters were included/excluded, but longer term, I'm hoping to better surface this information in a dedicated pane or something similar.

I think this approach makes sense for the Desktop user case, but let me think on the admin situation a bit! I feel like we will want a separate flag/option elsewhere, like an "admin policy" to enforce certain options

sharon-wang committed:
add js user specified locator unit test

added a comment to the native python finder to add a test in the future

sharon-wang committed:
convert aliased paths to absolute paths

sharon-wang committed:
💅 run prettier

seeM reviewed (commented):
Code changes LGTM! I tested locally and it seems to work well, except that the interpreters are still shown in the "Python: Select Interpreter" command. Is that expected?

sharon-wang committed:
wrap setInterpreter quickpick filter to also filter by user settings

sharon-wang committed:
format and comment fixes

sharon-wang committed:
traceInfo ignored relative paths instead of traceVerbose

isabelizimm commented on extensions/positron-python/src/client/pythonEnvironments/base/locators/common/nativePythonFinder.ts:
```diff
@@ -21,7 +21,11 @@ import { NativePythonEnvironmentKind } from './nativePythonUtils';
 import { StopWatch } from '../../../../common/utils/stopWatch';
 import { untildify } from '../../../../common/helpers';
-import { traceError } from '../../../../logging';
+import { traceError, traceVerbose } from '../../../../logging';
```
nit, put this in `// --- Start Positron ---`

sharon-wang commented on extensions/positron-python/src/client/interpreter/configuration/interpreterSelector/commands/setInterpreter.ts:
```diff
@@ -719,3 +728,15 @@ function getGroup(item: IInterpreterQuickPickItem, workspacePath?: string) {
             return EnvGroups[item.interpreter.envType];
     }
 }
+
+// --- Start Positron ---
+/**
+ * Wraps the original filter function to include additional filtering logic.
+ * If no filter is provided, the returned function will only include the additional filtering logic.
+ * @param filter The original filter function
+ * @returns A new filter function that includes the original filter function and the additional filtering logic
+ */
+function filterWrapper(filter: ((i: PythonEnvironment) => boolean) | undefined) {
+    return (i: PythonEnvironment) => (filter ? filter(i) : true) && shouldIncludeInterpreter(i.path);
+}
+// --- End Positron ---
```
@seeM
> the interpreters are still shown in the "Python: Select Interpreter" command. Is that expected?

Thank you for catching this! I checked the "include" case, but missed the "exclude" case for the quickpick. This should be fixed now!

sharon-wang committed:
💅

isabelizimm commented on extensions/positron-python/src/test/pythonEnvironments/nativePythonFinder.unit.test.ts:
```diff
@@ -17,6 +17,7 @@ import * as workspaceApis from '../../client/common/vscodeApis/workspaceApis';
 
 // --- Start Positron ---
 // re-enable when Native Finder is used for discovery
+// TODO: add test for python.interpreters.include here once we switch to Native Finder
```
this is fine for now, I'll log an issue to turn these back on (probably for the next milestone)

isabelizimm reviewed (commented):
Question: If I have started an interpreter, say, a pyenv interpreter, then I exclude `~/.pyenv`, should I expect to see the interpreter on restart? 

It will not show if I have shut down the interpreter, but if I leave it running and then quit/reopen Positron, the interpreter persists. This feels like the right choice, but just making sure it is intentional!


++ to @seeM comment, this does not affect the `Python: Select Interpreter` list. With the upcoming changes to our interpreter, we might want to make sure include/exclude works there as well

isabelizimm commented:
oh, just a few mins too late! I see the `Python: Select Interpreter` changes 🚀

sharon-wang committed:
add mocks/stubs for setInterpreter test

sharon-wang committed:
move traceVerbose import into positron fence

isabelizimm reviewed (dismissed):
good with me once ci is green ✅

sharon-wang committed:
fix lint

sharon-wang commented:
@isabelizimm
> Question: If I have started an interpreter, say, a pyenv interpreter, then I exclude ~/.pyenv, should I expect to see the interpreter on restart?
> 
> It will not show if I have shut down the interpreter, but if I leave it running and then quit/reopen Positron, the interpreter persists. This feels like the right choice, but just making sure it is intentional!

Yes this feels right to me as well 👍 Thanks for calling this out! I added a note to the QA Notes section to mention this case

isabelizimm committed:
python: filter versions less than 3.8 from interpreter dropdown (#6433)

### Release Notes

<!--
Optionally, replace `N/A` with text to be included in the next release
notes.
The `N/A` bullets are ignored. If you refer to one or more Positron
issues,
these issues are used to collect information about the feature or
bugfix, such
as the relevant language pack as determined by Github labels of type
`lang: `.
  The note will automatically be tagged with the language.

These notes are typically filled by the Positron team. If you are an
external
  contributor, you may ignore this section.
-->

#### New Features

- N/A

#### Bug Fixes

- #3740 filters out unsupported (<3.8) Python versions from the `Python:
Select Interpreter` dropdown


### QA Notes

1. create a Python interpreter <3.8 (if you are a pyenv user, you can do
`pyenv install 3.6`)
2. go to `Python: Select Interpreter`
3. should not appear in dropdown

melissa-barca reviewed (approved):
LGTM!

melissa-barca reviewed (changes requested):
Rescinding my approval, because we talked offline and Sharon is going to make a few modificatons

sharon-wang committed:
add `python.interpreters.debugInfo` command

Opens up the Python Language Pack Output and prints debug info for Python interpreter settings and all discovered interpreters.

sharon-wang committed:
change precedence to excludes over includes

melissa-barca reviewed (approved):
LGTM!
